src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs

```diff
@@ -132,11 +140,18 @@ public BuildEventArgsWriter(BinaryWriter binaryWriter)
         /// </summary>
         public void Write(BuildEventArgs e)
         {
-            WriteCore(e);
+            BinaryLogRecordKind eventKind = WriteCore(e);
 
-            // flush the current record and clear the MemoryStream to prepare for next use
-            currentRecordStream.WriteTo(originalStream);
-            currentRecordStream.SetLength(0);
+            FlushRecordToFinalStream(eventKind, currentRecordStream);
+        }
+
+        private void FlushRecordToFinalStream(BinaryLogRecordKind recordKind, MemoryStream recordStream)
+        {
+            using var redirectionScope = RedirectWritesToOriginalWriter();
+            Write(recordKind);
+            Write((int)recordStream.Length);
+            recordStream.WriteTo(originalStream);
+            recordStream.SetLength(0);
```

Comment:
NIT: I think, that preparing stream for reuse is better to do on acquiring side, than on releasing side. If releaseing do not happen for future changes or bugs (conditional, exception, etc...) stream is in unknown state.
Since SetLength(0) is very fast operation, it would not harm to have it on both, acquiring and releasing side.